File: core/src/main/java/dev/bee/kanjianki/core/RetentionSettingsPolicy.java
package dev.bee.kanjianki.core;

public final class RetentionSettingsPolicy {
    public static final String SAVED_MESSAGE = "FSRS retention saved.";

    private RetentionSettingsPolicy() {
    }

    public static SaveResult saveRequest(
            int retentionPercent,
            boolean frequencyRetentionEnabled,
            String frequencyRetentionRanges,
            RecordsSchedulerModels.SchedulerParameters latest
    ) {
        String ranges = frequencyRetentionRanges == null ? "" : frequencyRetentionRanges.trim();
        if (frequencyRetentionEnabled) {
            try {
                FrequencyRetentionRanges.parse(ranges);
            } catch (IllegalArgumentException error) {
                return SaveResult.invalid(error.getMessage());
            }
        }
        RecordsSchedulerModels.SchedulerParameters safeLatest = latest == null
                ? RecordsSchedulerModels.SchedulerParameters.defaults()
                : latest;
        RecordsSchedulerModels.SchedulerParameters parameters = new RecordsSchedulerModels.SchedulerParameters(
                SettingsInputRules.retentionPercent(retentionPercent / 100.0) / 100.0,
                safeLatest.againMultiplier,
                safeLatest.hardMultiplier,
                safeLatest.goodMultiplier,
                safeLatest.easyMultiplier,
                safeLatest.lastAdjustedAtMillis,
                safeLatest.lastAdjustmentReviewCount
        ).withFrequencyRetention(frequencyRetentionEnabled, ranges);
        return SaveResult.valid(parameters);
    }

    public static final class SaveResult {
        public final boolean valid;
        public final RecordsSchedulerModels.SchedulerParameters parameters;
        public final String message;

        private SaveResult(boolean valid, RecordsSchedulerModels.SchedulerParameters parameters, String message) {
            this.valid = valid;
            this.parameters = parameters;
            this.message = message;
        }

        static SaveResult valid(RecordsSchedulerModels.SchedulerParameters parameters) {
            return new SaveResult(true, parameters, SAVED_MESSAGE);
        }

        static SaveResult invalid(String message) {
            return new SaveResult(false, null, message);
        }
    }
}

File: core/src/main/kotlin/dev/bee/kanjianki/core/RetentionSettingsPolicy.kt
package dev.bee.kanjianki.core

import java.util.Locale

object RetentionSettingsPolicy {
    const val SAVED_MESSAGE: String = "Review retention saved."

    private const val JAPANESE_LANGUAGE = "ja"

    @JvmStatic
    fun saveRequest(
        retentionPercent: Int,
        frequencyRetentionEnabled: Boolean,
        frequencyRetentionRanges: String?,
        latest: RecordsSchedulerModels.SchedulerParameters?,
    ): SaveResult {
        val ranges = frequencyRetentionRanges?.trim().orEmpty()
        if (frequencyRetentionEnabled) {
            try {
                FrequencyRetentionRanges.parse(ranges)
            } catch (error: IllegalArgumentException) {
                return SaveResult.invalid(error.message)
            }
        }
        val safeLatest = latest ?: RecordsSchedulerModels.SchedulerParameters.defaults()
        val parameters = safeLatest
            .withTargetRetention(SettingsInputRules.retentionPercent(retentionPercent / 100.0) / 100.0)
            .withFrequencyRetention(frequencyRetentionEnabled, ranges)
        return SaveResult.valid(parameters)
    }

    @JvmStatic
    fun savedMessage(): String = localizedText(SAVED_MESSAGE, "レビュー保持率を保存しました。")

    private fun localizedText(english: String, japanese: String): String =
        if (isJapaneseLocale()) japanese else english

    private fun isJapaneseLocale(): Boolean = Locale.getDefault().language == JAPANESE_LANGUAGE

    class SaveResult private constructor(
        @JvmField val valid: Boolean,
        @JvmField val parameters: RecordsSchedulerModels.SchedulerParameters?,
        @JvmField val message: String?,
    ) {
        companion object {
            fun valid(parameters: RecordsSchedulerModels.SchedulerParameters): SaveResult {
                return SaveResult(true, parameters, RetentionSettingsPolicy.savedMessage())
            }

            fun invalid(message: String?): SaveResult {
                return SaveResult(false, null, message)
            }
        }
    }
}
